File: android/src/main/java/expo/modules/helpscout/ExpoHelpscoutModule.kt
package expo.modules.helpscout

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.URL
import android.text.TextUtils
import android.util.Log
import android.content.Context
import com.helpscout.beacon.Beacon
import com.helpscout.beacon.ui.BeaconActivity
import com.helpscout.beacon.model.BeaconScreens
import com.helpscout.beacon.model.SuggestedArticle


class ExpoHelpscoutModule : Module() {
  private val TAG = "com.driversnote.helpscoutbeacon"
  private var beacon: Beacon? = null

  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoHelpscout')` in JavaScript.
    Name("ExpoHelpscout")

    Function("init") { beaconID: String? ->
      if (beaconID.isNullOrEmpty()) {
        Log.w(TAG, "[init] Missing argument: beaconID")
        return@Function
      }
      beacon = Beacon.Builder()
        .withBeaconId(beaconID)
        .build()
    }

    Function("identify") { email: String?, name: String? ->
      if (beacon == null || email.isNullOrEmpty() || name.isNullOrEmpty()) {
        if (beacon == null) Log.w(TAG, "[identifyWithEmailAndName] Not initialized - did you forget to call 'init'?")
        if (email.isNullOrEmpty()) Log.w(TAG, "[identifyWithEmailAndName] Missing argument: email")
        if (name.isNullOrEmpty()) Log.w(TAG, "[identifyWithEmailAndName] Missing argument: name")
        return@Function
      }
      Beacon.identify(email, name)
    }

    Function("logout") {
      if (beacon == null) {
        Log.w(TAG, "[logout] Not initialized - did you forget to call 'init'?")
        return@Function null
      }
      Beacon.logout()
      null
    }

    Function("addAttributeWithKey") { key: String?, value: String? ->
      if (beacon == null || key.isNullOrEmpty() || value.isNullOrEmpty()) {
        if (beacon == null) Log.w(TAG, "[addAttributeWithKey] Not initialized - did you forget to call 'init'?")
        if (key.isNullOrEmpty()) Log.w(TAG, "[addAttributeWithKey] Missing argument: key")
        if (value.isNullOrEmpty()) Log.w(TAG, "[addAttributeWithKey] Missing argument: value")
        return@Function
      }
      Beacon.addAttributeWithKey(key, value)
    }

    Function("open") { signature: String? ->
      if (beacon == null) {
        Log.w(TAG, "[open] Not initialized - did you forget to call 'init'?")
        return@Function
      }
      val context = appContext.reactContext ?: return@Function
      if (signature.isNullOrEmpty()) {
        BeaconActivity.open(context)
      } else {
        BeaconActivity.openInSecureMode(context, signature)
      }
    }
//
    Function("openArticle") { articleID: String?, signature: String? ->
      if (beacon == null || articleID.isNullOrEmpty()) {
        if (beacon == null) Log.w(TAG, "[openArticle] Not initialized - did you forget to call 'init'?")
        if (articleID.isNullOrEmpty()) Log.w(TAG, "[openArticle] Missing argument: articleID")
        return@Function
      }

      val context = appContext.reactContext ?: return@Function
      val articleList = arrayListOf(articleID)

      if (signature.isNullOrEmpty()) {
        BeaconActivity.open(context, BeaconScreens.ARTICLE_SCREEN, articleList)
      } else {
        BeaconActivity.openInSecureMode(context, signature, BeaconScreens.ARTICLE_SCREEN, articleList)
      }
    }

    Function("navigate") { path: String? ->
      if (beacon == null) {
        Log.w(TAG, "[navigate] Not initialized - did you forget to call 'init'?")
        return@Function
      }

      val context = appContext.reactContext ?: return@Function
      when (path) {
        "/ask/message/" -> BeaconActivity.open(context, BeaconScreens.CONTACT_FORM_SCREEN, arrayListOf())
        "/ask/chat/" -> BeaconActivity.open(context, BeaconScreens.CHAT, arrayListOf())
        "/answers/" -> BeaconActivity.open(context, BeaconScreens.PREVIOUS_MESSAGES, arrayListOf())
        else -> Log.w(TAG, "[navigate] Path '${path ?: "null"}' not supported")
      }
    }

    Function("suggestArticles") { articleIds: List<String>? ->
      if (beacon == null) {
        Log.w(TAG, "[suggestArticles] Not initialized - did you forget to call 'init'?")
        return@Function
      }

      if (articleIds.isNullOrEmpty()) {
        Log.w(TAG, "[suggestArticles] Missing or empty articleIds")
        return@Function
      }

      val suggestedArticles = articleIds.take(5).map {
        SuggestedArticle.SuggestedArticleWithId(it)
      }

      Beacon.setOverrideSuggestedArticlesOrLinks(suggestedArticles)
    }

    Function("resetSuggestions") {
      if (beacon == null) {
        Log.w(TAG, "[resetSuggestions] Not initialized - did you forget to call 'init'?")
        return@Function null
      }
      Beacon.setOverrideSuggestedArticlesOrLinks(emptyList())
    }

  }
}

File: ios/ExpoHelpscoutModule.swift
import ExpoModulesCore
import Beacon

public class ExpoHelpscoutModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoHelpscout')` in JavaScript.
    Name("ExpoHelpscout")

      var helpscoutBeaconID: String? = nil
    var beaconUser: HSBeaconUser? = nil

    Function("init") { (beaconID: String?) in
      guard let beaconID = beaconID else {
        print("[init] missing parameter: beaconID")
        return
      }
      
      helpscoutBeaconID = beaconID
    }

    Function("identify") { (email: String?, name: String?) in
      guard let beaconID = helpscoutBeaconID else {
        print("[identify] Not initialized - did you forget to call 'init'?")
        return
      }

      guard let email = email else {
        print("[identify] missing parameter: email")
        return
      }

      let user = HSBeaconUser()
      user.email = email
      if let name = name {
        user.name = name
      }

      beaconUser = user
      HSBeacon.login(user)
    }


    Function("addAttributeWithKey") { (key: String?, value: String?) in
      guard let beaconID = helpscoutBeaconID else {
        print("[addAttributeWithKey] Not initialized - did you forget to call 'init'?")
        return
      }

      guard let user = beaconUser else {
        print("[addAttributeWithKey] Not initialized - did you forget to call 'identify' or 'login'?")
        return
      }

      guard let key = key, let value = value else {
        print("[addAttributeWithKey] missing parameters")
        return
      }

      user.addAttribute(withKey: key, value: value)
    }

    Function("open") { (signatureKey: String?) in
      guard let beaconID = helpscoutBeaconID else {
        print("[open] Not initialized - did you forget to call 'init'?")
        return
      }

      guard let _ = beaconUser else {
        print("[open] Not initialized - did you forget to call 'identify' or 'login'?")
        return
      }

      let settings = HSBeaconSettings(beaconId: beaconID)
      DispatchQueue.main.async {
        if let signature = signatureKey {
          HSBeacon.open(settings, signature: signature)
        } else {
          HSBeacon.open(settings)
        }
      }
    }


    Function("navigate") { (path: String?) in
      guard let beaconID = helpscoutBeaconID else {
        print("[navigate] Not initialized - did you forget to call 'init'?")
        return
      }

      guard let path = path else {
        print("[navigate] missing parameter: path")
        return
      }

      let settings = HSBeaconSettings(beaconId: beaconID)
      DispatchQueue.main.async {
        HSBeacon.navigate(path, beaconSettings: settings)
      }

    // MARK: - logout()
    Function("logout") {
      guard let _ = helpscoutBeaconID else {
        print("[logout] Not initialized - did you forget to call 'init'?")
        return
      }

      HSBeacon.logout()
    }

    // MARK: - openArticle(articleID, signature)
    Function("openArticle") { (articleID: String?, signature: String?) in
      guard let beaconID = helpscoutBeaconID else {
        print("[openArticle] Not initialized - did you forget to call 'init'?")
        return
      }

      guard let articleID = articleID else {
        print("[openArticle] missing parameter: articleID")
        return
      }

      let settings = HSBeaconSettings(beaconId: beaconID)
      DispatchQueue.main.async {
        if let signature = signature {
          HSBeacon.openArticle(articleID, beaconSettings: settings, signature: signature)
        } else {
          HSBeacon.openArticle(articleID, beaconSettings: settings)
        }
      }
    }

    Function("suggestArticles") { (articleIDList: [String]?) in
      guard let _ = helpscoutBeaconID else {
        print("[suggestArticles] Not initialized - did you forget to call 'init'?")
        return
      }

      guard let ids = articleIDList else {
        print("[suggestArticles] missing parameter: articleIDList")
        return
      }

      HSBeacon.suggest(ids)
    }

    Function("resetSuggestions") {
      guard let _ = helpscoutBeaconID else {
        print("[resetSuggestions] Not initialized - did you forget to call 'init'?")
        return
      }

      HSBeacon.suggest([])
    }
  }
}}
